File: algorithms/java/src/test/java/leetcode/LC264UglyNumberII.java
package leetcode;

import org.junit.Test;

import java.util.PriorityQueue;

import static org.junit.Assert.assertEquals;

/**
 * https://leetcode.com/problems/ugly-number-ii/
 * #medium
 */
public class LC264UglyNumberII {
    public class Solution {
        public int nthUglyNumber(int n) {
            if (n == 1) return 1;
            PriorityQueue<Long> uglies = new PriorityQueue();
            uglies.add(1l);
            for (int i = 1; i < n; i++) {
                long smallest = uglies.poll();
                while (!uglies.isEmpty() && uglies.peek() == smallest) {
                    uglies.poll();
                }
                uglies.add(smallest * 2);
                uglies.add(smallest * 3);
                uglies.add(smallest * 5);
            }
            return uglies.poll().intValue();
        }
    }

    @Test
    public void test_example() throws Exception {
        int[] expected = {1, 2, 3, 4, 5, 6, 8, 9, 10, 12};
        for (int i = 0; i < expected.length; i++) {
            assertEquals(expected[i], new Solution().nthUglyNumber(i + 1));
        }
    }

    @Test
    public void test_bigger() throws Exception {
        assertEquals(536870912, new Solution().nthUglyNumber(1407));
    }
}

File: algorithms/code/leetcode/lc264_ugly_number_ii/lc264_ugly_number_ii.py
#!/usr/bin/env python3
# https://leetcode.com/problems/super-ugly-number/

import heapq
import unittest


class Solution:
    def nthUglyNumber(self, n: int) -> int:
        if n == 1:
            return 1
        uglies = [1]
        for _ in range(1, n):
            smallest = heapq.heappop(uglies)
            while uglies and uglies[0] == smallest:
                heapq.heappop(uglies)
            heapq.heappush(uglies, smallest * 2)
            heapq.heappush(uglies, smallest * 3)
            heapq.heappush(uglies, smallest * 5)
        return uglies[0]


class TestCode(unittest.TestCase):
    def test_example(self):
        expected = [1, 2, 3, 4, 5, 6, 8, 9, 10, 12]
        for i, expected_i in enumerate(expected):
            self.assertEqual(expected_i, Solution().nthUglyNumber(i + 1))

    def test_bigger(self):
        self.assertEqual(536870912, Solution().nthUglyNumber(1407))
